File: Calculator/ios-native/CalculatorHub/Calculators/CompoundInterestCalculator.swift
import Foundation

struct CompoundInterestResult {
    let futureValue: Double
    let totalContributions: Double
    let totalInterest: Double
}

class CompoundInterestCalculator {
    func calculate(
        principal: Double,
        annualRate: Double,
        years: Double,
        contribution: Double = 0.0,
        compoundsPerYear: Double = 12.0
    ) -> CompoundInterestResult {
        let rate = annualRate / 100.0
        let n = compoundsPerYear
        
        // Future Value of Principal: P(1 + r/n)^(nt)
        let fvPrincipal = principal * pow(1 + rate / n, n * years)
        
        // Future Value of Series (Contributions): PMT * [((1 + r/n)^(nt) - 1) / (r/n)]
        let fvContributions: Double
        if rate > 0 {
            fvContributions = contribution * (pow(1 + rate / n, n * years) - 1) / (rate / n)
        } else {
            fvContributions = contribution * n * years
        }
        
        let total = fvPrincipal + fvContributions
        let totalContributed = principal + (contribution * n * years)
        let totalInterest = total - totalContributed
        
        return CompoundInterestResult(
            futureValue: total,
            totalContributions: totalContributed,
            totalInterest: totalInterest
        )
    }
}

File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/CompoundInterestCalculator.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.pow

data class CompoundInterestResult(
    val futureValue: Double,
    val totalContributions: Double,
    val totalInterest: Double
)

class CompoundInterestCalculator {
    fun calculate(
        principal: Double,
        annualRate: Double,
        years: Double,
        contribution: Double = 0.0,
        compoundsPerYear: Double = 12.0
    ): CompoundInterestResult {
        val rate = annualRate / 100.0
        val n = compoundsPerYear

        // Future Value of Principal: P(1 + r/n)^(nt)
        val fvPrincipal = principal * (1 + rate / n).pow(n * years)

        // Future Value of Series (Contributions): PMT * [((1 + r/n)^(nt) - 1) / (r/n)]
        val fvContributions = if (rate > 0) {
            contribution * ((1 + rate / n).pow(n * years) - 1) / (rate / n)
        } else {
            contribution * n * years
        }

        val total = fvPrincipal + fvContributions
        val totalContributed = principal + (contribution * n * years)
        val totalInterest = total - totalContributed

        return CompoundInterestResult(
            futureValue = total,
            totalContributions = totalContributed,
            totalInterest = totalInterest
        )
    }
}

